carla/recourse_methods/catalog/revise/model.py

```diff
@@ -0,0 +1,206 @@
+                    {
+                        "d": 8,  # latent space
+                        "D": test_factual.shape[1],  # input size
+                        "H1": 512,
+                        "H2": 256,
+                        "train": False,
+                        "lambda_reg": 1e-6,
+                        "epochs": 5,
+                        "lr": 1e-3,
+                        "batch_size": 32,
+                    }
+            }
+        """
+        super().__init__(mlmodel)
+        self.params = hyperparams
+
+        self._target_column = data.target
+        self._lambda = self.params["lambda"] if "lambda" in hyperparams.keys() else 0.5
+        self._optimizer = (
+            self.params["optimizer"] if "optimizer" in hyperparams.keys() else "adam"
+        )
+        self._lr = self.params["lr"] if "lr" in hyperparams.keys() else 0.1
+        self._max_iter = (
+            self.params["max_iter"] if "max_iter" in hyperparams.keys() else 1000
+        )
+        self._target_class = (
+            hyperparams["target_class"]
+            if "target_class" in hyperparams.keys()
+            else [0, 1]
+        )
+        self._binary_cat_features = (
+            hyperparams["binary_cat_features"]
+            if "binary_cat_features" in hyperparams.keys()
+            else True
+        )
+
+        df_enc_norm_data = self.encode_normalize_order_factuals(
+            data.raw, with_target=True
+        )
+
+        vae_params = hyperparams["vae_params"]
+        self.vae = VariationalAutoencoder(
+            self.params["data_name"],
+            vae_params["d"],
+            df_enc_norm_data.shape[1] - 1,  # num features - target
+            vae_params["H1"],
+            vae_params["H2"],
+        )
+
+        if vae_params["train"]:
+            self.vae = train_variational_autoencoder(
+                self.vae,
+                self._mlmodel.data,
+                self._mlmodel.scaler,
+                self._mlmodel.encoder,
+                self._mlmodel.feature_input_order,
+                lambda_reg=vae_params["lambda_reg"],
+                epochs=vae_params["epochs"],
+                lr=vae_params["lr"],
+                batch_size=vae_params["batch_size"],
+            )
+        else:
+            try:
+                self.vae.load(df_enc_norm_data.shape[1] - 1)
+            except FileNotFoundError as exc:
+                raise FileNotFoundError(
+                    "Loading of Autoencoder failed. {}".format(str(exc))
+                )
+
+    def get_counterfactuals(self, factuals: pd.DataFrame) -> pd.DataFrame:
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        df_enc_norm_fact = self.encode_normalize_order_factuals(
+            factuals, with_target=True
+        )
+
+        # prepare data for optimization steps
+        test_loader = torch.utils.data.DataLoader(
+            Dataloader(df_enc_norm_fact.values), batch_size=1, shuffle=False
```

Comment:
Pass a DataLoader into a DataLoader, thats how it supposed to be? 😁

Reply:
The inner DataLoader is more a custom dataset, inherited from torch Dataset :smile: 
Will rename it to make this clearer.